### ISSUE TITLE
Connectivity failure when `maker` , `taker` & `dns` runs on different network

### ISSUE THREAD

**Issue by KnowWhoami** [citadel-tech/coinswap#321]
Let us consider the  state when  `makerd` tries to connect to  `dns` server for posting its address the  to it's address book
Here's our current logic: 
https://github.com/citadel-tech/coinswap/blob/6e5ce00f710685cd43b44a78896b7d1a3c6c6648/src/maker/server.rs#L64-L136
 
 we can split it into two cases:
# Case 1: when both `dns`  & `makerd` run on same network: (For eg: `dns & makerd -> tor`)
- As per the logic , `makerd`  will consider the hardcoded dns address in the case of `IT` (Here , it will be `onion` address because **the makerd is running on `tor` not because of the `dns` network**) or directly take it as `directory_server_address` from `MakerConfig` itself.
- **Since the `makerd` is running on tor so it assumes that the `dns` also runs on `tor` and it's derived address is of `tor`.**
- Here, in this case it works since both runs on same network , similarly  happens when both runs on `clearnet` as well.
- But this will not work in case when both run on different network.

# Case 2: Both run on different network (For eg: `dns-> clearnet` & `makerd -> tor`)

- Similar to what mentioned above in case 1 -> Maker ,since running on tor  ->  abruptly assumes that the dns also works on tor and it's derived address is of `onion` one.
-  So as per the logic,  `makerd` would try to connect to dns (assumes it's run on tor) via `socks_proxy` but it would fail to connect as the dns is on clearnet.
- Similar behaviour can be seen when `maker` is on `clearnet` while `dns` on `tor`.

This bug can be observed when `taker` tries to connect to `maker` or `dns` as well.

**Comment by KnowWhoami** [comment]
> This is not a bug, we explicitly made it like this.

Why? What's the reason?
IMO, It will severly hinder the performance of whole Coinswap protocol in terms of privacy , flexibility  in choosing network to maker & taker as : 
- Let's say,  `Directory`  works on `clearnet` -> then both `maker` & `taker` will be forced to  run on `clearnet` as well for smooth functioning since they won't be able to connect to each  other in case of different network ->  If that's the case then what's the purpose of providing flexibility to run on any network -> after all , they have to work on particular one.
- And , they can't even run on `tor`, if they wish to  -> thus can't maintain privacy, etc.

Similar downsides can be seen when all works in `tor` as well.

**Comment by mojoX911** [comment]
> Why? What's the reason?

because it was easy to implement.

This is a problem and yes, both the networks together are not compatible with each other. 

There are two ways of solving this:

- Assume only one network. Always tor. So everybody is forced to be in tor only and clearnet address is strictly not allowed. Which isn't a very bad idea. running these type of servers in clearnet can have legal implications. This is simple to do in our current code, no refactorings are required. But Tor has to run consistently for all OS. Currently, we face an issue with running Tor in Mac. 

- Make all the systems, Maker, Taker and DNS to work on both networks in parallel. Maker should can give either/both of their tor and clearnet address to DNS. Taker should be able to send/recv via both tor and clearnet. Same for DNS. This is quite a large code. Will require refactoring of our networking code.

The 2nd way is more complete and can provide backup when tor fails, which is quite often. But in theory: "Nobody should ever run a maker server in clearnet".

The absolute best case would be, running 2 non-clearnet networks in parallel. Like Tor + I2P.

**Comment by mojoX911** [comment]
@Shourya742 @KnowWhoami @007harshmahajan can this be closed now as we decided to keep the network tor only for now?

**Comment by 007harshmahajan** [comment]
This can also be closed
